### PR TITLE
Handle fancy wildcard branch patterns

### PR DESCRIPTION
**Pull request by lox**
Based on feedback in https://forum.buildkite.community/t/cli-steps-dont-run-when-there-is-a-negated-branch-filter/400/5, this adds support for different branch filter patterns. 

@keithpitt I'd love some feedback on what the actual types of patterns that are supported are 😅

### CODE REVIEW

**Comment by keithpitt** [comment]
These are "matching" fixtures. Left is the pattern, right is the branch. All return "yes, it's a match"

```ruby
# empty branch names
[ "foo", nil ],
[ "", nil ],
[ "    ", nil ],

# empty branch configs
[ nil, "foo" ],
[ nil, "" ],
[ nil, "       " ],

# single name branch matches
[ "foo bar", "foo" ],
[ "foo             bar", "bar" ],

# with whitespace
[ "    master   ", "master" ],
[ "master   ", "master" ],
[ "   master", "master" ],

# ones with a slash
[ "feature/authentication", "feature/authentication" ],

# not-checking
[ "!foo", "master" ],
[ "!release/production !release/test", "master" ],

# prefix wildcards
[ "*-do-the-thing", "can-you-do-the-thing" ],
[ "!*-do-the-thing", "can-you-do-the-thing-please" ],

# wildcards
[ "release/*", "release/foo" ],
[ "release/*", "release/bar/bing" ],
[ "release-*", "release-thingy" ],
[ "release-* release/*", "release-thingy" ],
[ "release-* release/*", "release/thingy" ],
[ "this-*-thing-is-the-*", "this-ruby-thing-is-the-worst" ],
[ "this-*-thing-is-the-*", "this-regex-thing-is-the-best" ],
[ "this-*-thing-is-the-*", "this-*-thing-is-the-*" ],
[ "this-*-thing-is-the-*", "this--thing-is-the-best-" ],
```

Here's are the ones that return `false`, not a match:

```ruby
# branch names
[ "foo         bar", "bang" ],

# not-matchers
[ "!foo bar", "foo" ],
[ "!release/*", "release/foo" ],
[ "!release/*", "release/bar" ],
[ "!refs/tags/*", "refs/tags/production" ],
[ "!release/production !release/test", "release/production" ],
[ "!release/production !release/test", "release/test" ],

# ones with a slash
[ "feature/authentication", "update/deployment" ],

# wildcards
[ "release-*", "release/thingy" ],
[ "release-*", "master" ],
[ "*-do-the-thing", "this-is-not-the-thing" ],
```

**Comment by keithpitt** [comment]
Feel free to pilfer the data and add them into the test suite!

File: local/pipeline.go [hunk 1]
```diff
@@ -42,15 +42,23 @@ func (s *step) UnmarshalJSON(data []byte) error {
 		branches = b
 	}
 
-	// Handle various types of branch vs branches
 	if branches != nil {
+		// Sometimes branches are space delimited, sometimes comma
```
Comment (keithpitt) [inline local/pipeline.go]: BK backend doesn't support `branches: "foo, bar"` - is that what this bit is for? Or is this something else?

Comment (lox) [reply local/pipeline.go]: Ok cool, not sure where I got that from, but will ditch it.

**Comment by lox** [comment]
Mind reviewing where I got up to when you get a minute @keithpitt?

**Comment by lox** [comment]
Gonna yolo merge this.

**Comment by gerrywastaken** [comment]
Thanks @lox!

**Pull request merged**